artmanflow/steps/common.py

```diff
@@ -0,0 +1,367 @@
+        with open(input_file_name, 'r') as input_file, \
+            open(output_file_name, 'a+') as output_file:
+            output_file.write(
+                "\033[1;30msubprocess.Popen(%s, stdout='%s', stderr='%s', stdin='%s'\n\033[0m\n" % (
+                    cmd, output_file_name, output_file_name, input_file_name))
+            output_file.flush()
+            subprocess.Popen(cmd, stdout=output_file, stderr=output_file,
+                             stdin=input_file)
+
+    def _construct_docker_run_command(self, command, guest_root_path=None,
+        mount_dirs=None):
+        cmd = ['docker', 'run', '--rm']  # run and remove container after exit
+        if guest_root_path:
+            cmd += ['-w', guest_root_path]  # set working dir
+        if mount_dirs:
+            for mount_dirs_pair in mount_dirs:
+                cmd += ['-v', ':'.join(mount_dirs_pair)]  # mount volumes
+        cmd += [
+            '-e', 'HOST_USER_ID=%s' % os.getuid(),  # to chown inside  docker
+            '-e', 'HOST_GROUP_ID=%s' % os.getgid(),
+            '-i', self._config['docker_image']  # specify image
+        ]
+        cmd += command
+
+        return cmd
+
+    def local_repo_mount(self, config):
+        repo = config['git_repo']
+        if repo.startswith('file://'):
+            local_mount_path = repo[len('file://'):]
+        elif repo.startswith('/'):
+            local_mount_path = repo
+        else:
+            return []
+
+        local_repo_name = local_mount_path[local_mount_path.rfind('/') + 1:]
+        guest_mount_path = self._guest.guest_root_subpath(local_repo_name)
+
+        return [local_mount_path, guest_mount_path]
+
+
+class BaseGuest:
+    def __init__(self, config):
+        self._config = config
+        self._guest = GuestStepProperties(config['guest_root_path'])
+
+    def before_execute(self):
+        self.puts(">>>>>>>>>> START GUEST SCRIPT EXECUTION: %s\n" % ' '.join(
+            sys.argv))
+        self.puts('Guest Script config:\n', '\033[1;30m', '\033[36m')
+        ConfigUtils.dump_config(self._config, sys.stdout)
+        self.puts(end_color='\033[0m')
+
+    def after_execute(self, save_guest_files_snapshot, exception=None):
+        snapshot_path = self._guest.guest_root_dir_snapshot_path()
+        if exception:
+            self.puts(
+                "Exception was thrown: %s. Saving guest script root"
+                " directory snapshot for debugging at: %s" % (
+                    str(exception), snapshot_path))
+        if exception or save_guest_files_snapshot:
+            self._create_guest_files_snapshot(snapshot_path)
+        self.puts("<<<<<<<<< END GUEST SCRIPT EXECUTION")
+
+    def puts(self, message='', start_color="\033[1;30m", end_color='\033[0m'):
+        sys.stdout.flush()
+        print("%s%s%s" % (start_color, message, end_color))
+        sys.stdout.flush()
+
+    def run_command(self, command, cwd=None, hide_command=False):
```

Comment:
We already have this function in artman; why do we need a copy of it?

Reply:
Please check the previous comment.